File: src/domain/repositories/customer.repository.ts
import { ICustomer } from '@domain/models';
import { Injectable } from '@nestjs/common';

@Injectable()
export class CustomerRepository {
  private customers: ICustomer[] = [];

  create(customer: ICustomer): Promise<ICustomer> {
    return new Promise((resolve) => {
      customer.id = (this.customers.length + 1).toString();
      customer.createAt = new Date();
      this.customers.push(customer);
      resolve(customer);
    });
  }

  update(customer: ICustomer): Promise<ICustomer> {
    return new Promise((resolve) => {
      const { id } = customer;
      customer.updateAt = new Date();
      const entity = this.customers.find((c) => c.id === id);
      Object.assign(entity, customer);

      resolve(customer);
    });
  }

  delete(id: string): Promise<void> {
    return new Promise((resolve) => {
      const index = this.customers.findIndex((c) => c.id === id);
      this.customers.splice(index, 1);
      resolve();
    });
  }

  findAll(): Promise<ICustomer[]> {
    return new Promise((resolve) => {
      resolve(this.customers);
    });
  }

  findById(id: string): Promise<ICustomer> {
    return new Promise((resolve) => {
      const customer = this.customers.find((c) => c.id === id);
      resolve(customer);
    });
  }
}

File: dotnet/Domain/Repositories/CustomerRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using dotnet.Domain.Models;

namespace dotnet.Domain.Repositories
{
    public class CustomerRepository : ICustomerRepository
    {
        private readonly MyAppDbContext _db;
        public CustomerRepository(MyAppDbContext db)
        {
            _db = db;
        }

        public Customer Create(Customer customer)
        {
            customer.Id = Guid.NewGuid();
            customer.CreatedAt = DateTime.Now;
            _db.Customers.Add(customer);
            _db.SaveChanges();

            return customer;
        }

        public Customer Update(Customer customer)
        {
            customer.UpdatedAt = DateTime.Now;
            _db.SaveChanges();
            return customer;
        }

        public void Delete(Customer customer)
        {
            _db.Customers.Remove(customer);
            _db.SaveChanges();
        }

        public List<Customer> FindAll()
        {
            return _db.Customers.ToList();
        }

        public Customer FindById(Guid id)
        {
            return _db.Customers.SingleOrDefault(q => q.Id == id);
        }
    }
}